plane/src/proxy/rewriter.rs

```diff
@@ -61,6 +62,19 @@ impl RequestRewriter {
         &self.bearer_token
     }
 
+    pub fn get_host_header(&self) -> Option<&HeaderValue> {
+        self.parts.headers.get(HOST)
+    }
+
+    pub fn get_subdomain(&self) -> Result<Option<&str>, ToStrError> {
+        Ok(self
+            .get_host_header()
+            .map(|h| h.to_str())
+            .transpose()?
+            .and_then(|h| h.rsplit_once('.'))
+            .map(|(before_domain_part, _)| before_domain_part))
+    }
```

Comment:
I think we want to thread through the cluster name so that we have it here, and then remove it as a suffix and return whatever's left. Otherwise, if we have a cluster like `p.jamsocket.net` without a subdomain this will still return `Ok(Some("p"))`.

Reply:
Fixed! Tried to leverage idiomatic Rust as much as possible -- let me know if you have any feedback! 

Setup:
```diff
$ git diff dev/
diff --git a/dev/controller.sh b/dev/controller.sh
index 6b471c3..f543f4f 100755
--- a/dev/controller.sh
+++ b/dev/controller.sh
@@ -3,5 +3,5 @@
 cargo run -- \
     controller \
     --db postgres://postgres@localhost \
-    --default-cluster localhost:9090 \
+    --default-cluster p.localhost:9090 \
     "$@"
diff --git a/dev/drone.sh b/dev/drone.sh
index 89c902a..a99c899 100755
--- a/dev/drone.sh
+++ b/dev/drone.sh
@@ -3,6 +3,6 @@
 cargo run -- \
     drone \
     --controller-url ws://localhost:8080/ \
-    --cluster "localhost:9090" \
+    --cluster "p.localhost:9090" \
     --db "drone.sqlite" \
     "$@"
diff --git a/dev/proxy.sh b/dev/proxy.sh
index 48b50bf..db51fc1 100755
--- a/dev/proxy.sh
+++ b/dev/proxy.sh
@@ -3,5 +3,5 @@
 cargo run -- \
     proxy \
     --controller-url ws://localhost:8080/ \
-    --cluster "localhost:9090" \
+    --cluster "p.localhost:9090" \
     "$@"
```

Demo:
```shell
$ ./dev/cli.sh connect --cluster 'p.localhost:9090' --image 'ghcr.io/drifting-in-space/demo-image-drop-four' --subdomain 'subdomain'
Created backend: qevn9013knwmgo
URL: http://subdomain.p.localhost:9090/btaUziMNMxuO-ujV7TwyxfrAb7uoJPOzSrQewdYNjjk/

$ curl -Is http://subdomain.p.localhost:9090/btaUziMNMxuO-ujV7TwyxfrAb7uoJPOzSrQewdYNjjk/
HTTP/1.1 200 OK
$ curl -Is http://p.localhost:9090/btaUziMNMxuO-ujV7TwyxfrAb7uoJPOzSrQewdYNjjk/
HTTP/1.1 401 Unauthorized
$ curl -Is http://subdomainp.localhost:9090/btaUziMNMxuO-ujV7TwyxfrAb7uoJPOzSrQewdYNjjk/
HTTP/1.1 400 Bad Request
$ curl -Is http://localhost:9090/btaUziMNMxuO-ujV7TwyxfrAb7uoJPOzSrQewdYNjjk/
HTTP/1.1 400 Bad Request
$ curl -Is http://subdomain.localhost:9090/btaUziMNMxuO-ujV7TwyxfrAb7uoJPOzSrQewdYNjjk/
HTTP/1.1 400 Bad Request

$ ./dev/cli.sh connect --cluster 'p.localhost:9090' --image 'ghcr.io/drifting-in-space/demo-image-drop-four'
Created backend: vsyu98ie6eud66
URL: http://p.localhost:9090/vQHaeFhv_OV6b-R0u0xAN5hMpCRTdYPbKS4Zozc9aUs/

$ curl -Is http://p.localhost:9090/vQHaeFhv_OV6b-R0u0xAN5hMpCRTdYPbKS4Zozc9aUs/
HTTP/1.1 200 OK
$ curl -Is http://sub.p.localhost:9090/vQHaeFhv_OV6b-R0u0xAN5hMpCRTdYPbKS4Zozc9aUs/
HTTP/1.1 401 Unauthorized
$ curl -Is http://subp.localhost:9090/vQHaeFhv_OV6b-R0u0xAN5hMpCRTdYPbKS4Zozc9aUs/
HTTP/1.1 400 Bad Request